src/core/game.js

```diff
@@ -0,0 +1,99 @@
+import Coords from './coords';
+
+export const EMPTY_VALUE = 0;
+
+export const buildGrid = (size = 4) => {
+    let value = EMPTY_VALUE;
+    let grid = [];
+
+    for (let y = 0; y < size; y++) {
+        grid[y] = [];
+
+        for (let x = 0; x < size; x++) {
+            value++;
+            grid[y][x] =
+                value == size * size ? EMPTY_VALUE : (grid[y][x] = value);
+        }
+    }
+    return grid;
+};
+
+export const deepCopyGrid = grid => grid.map(row => row.slice());
+
+export const areGridsEquals = (grid1, grid2) => {
+    if (!grid1 || !grid2) {
+        return false;
+    }
+
+    if (grid1.length != grid2.length) {
+        return false;
+    }
+    for (let i = 0; i < grid1.length; i++) {
```

Comment:
Recursive calls are hard to read. You could probably avoid it by using some `reduce` and/or `forEach` calls. (Same applies for everywhere there is a for loop for rows.